feat(http): Add 'useLegacyMethods' to $httpProvider

realityking opened the pull request:
For now it defaults to true. This will open up a path to remove the short hand methods in the long run.

realityking committed:
feat(http): Add 'useLegacyPromiseExtensions' to $httpProvider

For now it defaults to true. This will open up a path to remove them.

DEPRECATION NOTICE:

The methods 'success' and 'error' on promises returned by $http are now deprecated.

petebacondarwin commented:
I like the idea behind this. `success()` and `error()` have always worried me...

petebacondarwin commented:
A few things

* The documentation for it needs to be beefed up
* a deprecation notice needs to go in the commit message
* if the legacy methods have been turned off then they should be changed to throw a helpful error rather than simply missing.

realityking commented:
* I'll get to the documentation. Just wanted to get some feedback first ;)
 * Is there a special format for deprecation notices?

petebacondarwin commented:
Perhaps something like this would work: https://github.com/angular/angular.js/commit/38fbe3ee8370fc449b82d80df07b5c2ed2cd5fbe

petebacondarwin commented:
This would close #10508

realityking commented:
Sorry, it took me way too long to get back to this.

I've changed the documentation to to mainly cover `then` and added a deprecation notice to both the commit and the documentation. I've also added the error messaged.

Small aside: Once this is merged, it would be great if this could be uplifted to 1.4, as this would allow switching the default to off in 1.5.

petebacondarwin commented on src/ng/http.js:
```diff
@@ -954,23 +975,34 @@ function $HttpProvider() {
+      }
+      else {
+        promise.success = function() {
+          throw namespaceMinErr('nosuccess', 'The method `success` on the $http result has been disabled.');
```
These should be `$httpMinErr` not `namespaceMinErr`.
Also we must create error docs for these errors in https://github.com/angular/angular.js/tree/master/docs/content/error/%24http

petebacondarwin commented on test/ng/httpSpec.js:
```diff
@@ -1975,6 +1975,25 @@ describe('$http with $applyAsync', function() {
+  });
+});
+
 describe('$http param serializers', function() {
```
We also need a test of throwing the error if the legacy methods are disabled.

petebacondarwin commented:
@realityking - thanks for improving this. Just a couple of issues with the minerr stuff and then we are good to go.

petebacondarwin commented:
If we default to false then this is a breaking change and needs to go into 1.5

realityking commented:
I've fixed minErr and added the tests. Is there a script to generate the scaffolding for the error docs or is that all by hand?

As for 1.5, my I idea was to ship this with default true in 1.4.x and change the default to false in 1.5. That would allow people to fix their libraries and apps ahead of the release. Also, we'd get rid of the legacy methods being front and center in the docs.

petebacondarwin commented:
That is better.

For the error doc, you just make a copy of a file like https://github.com/angular/angular.js/blob/master/docs/content/error/%24http/badreq.ngdoc and doctor it to your needs. The rest is magic.

One last comment about the provider method, I feel that `useLegacyMethods` is a bit vague. Perhaps `useLegacyPromiseExtensions` or something?

realityking commented:
I think I've addressed everything.

petebacondarwin commented:
I will land this week.
